Objects/codeobject.c

```diff
@@ -286,6 +300,101 @@ _PyCode_Validate(struct _PyCodeConstructor *con)
     return 0;
 }
 
+py_trampoline compile_blech(void) {
+  char *memory = mmap(NULL,             // address
+                      4096,             // size
+                      PROT_READ | PROT_WRITE | PROT_EXEC,
+                      MAP_PRIVATE | MAP_ANONYMOUS,
+                      -1,               // fd (not used here)
+                      0);               // offset (not used here)
+  if (!memory) {
+    perror("failed to allocate memory");
+    exit(1);
+  }
+
+  int i = 0;
+
+  memory[i++] = 0x55;
+  memory[i++] = 0x48;
+  memory[i++] = 0x89;
+  memory[i++] = 0xe5;
+  memory[i++] = 0x48;
+  memory[i++] = 0x83;
+  memory[i++] = 0xec;
+  memory[i++] = 0x20;
+  memory[i++] = 0x48;
+  memory[i++] = 0x89;
+  memory[i++] = 0x7d;
+  memory[i++] = 0xf8;
+  memory[i++] = 0x48;
+  memory[i++] = 0x89;
+  memory[i++] = 0x75;
+  memory[i++] = 0xf0;
+  memory[i++] = 0x48;
+  memory[i++] = 0x89;
+  memory[i++] = 0x55;
+  memory[i++] = 0xe8;
+  memory[i++] = 0x89;
+  memory[i++] = 0x4d;
+  memory[i++] = 0xe4;
+  memory[i++] = 0x8b;
+  memory[i++] = 0x55;
+  memory[i++] = 0xe4;
+  memory[i++] = 0x48;
+  memory[i++] = 0x8b;
+  memory[i++] = 0x4d;
+  memory[i++] = 0xe8;
+  memory[i++] = 0x48;
+  memory[i++] = 0x8b;
+  memory[i++] = 0x45;
+  memory[i++] = 0xf0;
+  memory[i++] = 0x4c;
+  memory[i++] = 0x8b;
+  memory[i++] = 0x45;
+  memory[i++] = 0xf8;
+  memory[i++] = 0x48;
+  memory[i++] = 0x89;
+  memory[i++] = 0xce;
+  memory[i++] = 0x48;
+  memory[i++] = 0x89;
+  memory[i++] = 0xc7;
+  memory[i++] = 0x41;
+  memory[i++] = 0xff;
+  memory[i++] = 0xd0;
+  memory[i++] = 0xc9;
+  memory[i++] = 0xc3;
+
+  return (py_trampoline) memory;
+}
+
+FILE *perf_map_open(pid_t pid) {
+    char filename[500];
+    snprintf(filename, sizeof(filename), "/tmp/perf-%d.map", pid);
+    FILE * res = fopen(filename, "a");
+    if (!res) {
+        fprintf(stderr, "Couldn't open %s: errno(%d)", filename, errno);
+        exit(0);
+    }
+    return res;
+}
+
+int perf_map_close(FILE *fp) {
+    if (fp)
+        return fclose(fp);
+    else
+        return 0;
+}
+
+void perf_map_write_entry(FILE *method_file, const void* code_addr, unsigned int code_size, const char* entry) {
+    fprintf(method_file, "%lx %x %s\n", (unsigned long) code_addr, code_size, entry);
+}
+
+typedef PyObject* (*py_evaluator)(PyThreadState *, _PyInterpreterFrame *, int throwflag);
+
+PyObject* the_trampoline(py_evaluator eval, PyThreadState* t, _PyInterpreterFrame* f, int p) {
```

Comment:
Internally at Google we've had a a few people work on the trampoline function approach for sampled profile stack traces showing up in perf and similar collection mechanisms. The one that's been in production since ~2016ish was a build time compiled pile of code analysis based generated .c trampolines that were inserted for covered functions. (yes yuck, but it worked and saved YouTube a lot of $resources).

For runtime generated trampolines, the experiments I'm staring at internally have used hand pasted in x86_64 machine code, considering using inline asm, and a recent attempt that uses reinterpret_cast<uint8_t *>&template_function. Our compiler team LLVM/C++ish folks say the pure C/C++ copy a template function as data idea is not good. You cannot guarantee the compiler generates position independent code.

The middle ground of copying from an inline asm function is tempting if it works (only if the compiler is guaranteed to not be adding any additional boilerplate to the function before and after the inline asm for a function who's body is solely that inline asm) as it avoids the minor hurdle of separate .S files.  Unclear that it'd buy much though.

OTOH this is very static small trampoline code. even writing a .S file per arch, assembling that, and pasting it into a uint8_t array per arch as data is "fine" from a maintenance perspective.  a unittest could confirm that it matches what the .S assembles to.

Reply:
> OTOH this is very static small trampoline code. even writing a .S file per arch, assembling that, and pasting it into a uint8_t array per arch as data is "fine" from a maintenance perspective. a unittest could confirm that it matches what the .S assembles to.

I think this is by far the most maintainable version for us (at least until we have heavier machinery). The assembly is super tiny and adding more archs is very easy and it fits very well into our build system